File: arka-sdk-dotnet/Pact.Sdk/Condition.cs
using System.Text.Json.Serialization;

namespace Pact.Sdk;

/// <summary>
/// Base class for rule conditions.
/// </summary>
[JsonPolymorphic(TypeDiscriminatorPropertyName = "type")]
[JsonDerivedType(typeof(CompareCondition), "compare")]
[JsonDerivedType(typeof(AndCondition), "and")]
[JsonDerivedType(typeof(OrCondition), "or")]
[JsonDerivedType(typeof(NotCondition), "not")]
[JsonDerivedType(typeof(ExistsCondition), "exists")]
[JsonDerivedType(typeof(InCondition), "in")]
[JsonDerivedType(typeof(RangeCondition), "range")]
[JsonDerivedType(typeof(ExpressionCondition), "expression")]
public abstract record Condition;

/// <summary>
/// Comparison condition.
/// </summary>
public record CompareCondition : Condition
{
    [JsonPropertyName("field")]
    public required string Field { get; init; }

    [JsonPropertyName("operator")]
    public required string Operator { get; init; }

    [JsonPropertyName("value")]
    public required object Value { get; init; }
}

/// <summary>
/// Logical AND condition.
/// </summary>
public record AndCondition : Condition
{
    [JsonPropertyName("conditions")]
    public required List<Condition> Conditions { get; init; }
}

/// <summary>
/// Logical OR condition.
/// </summary>
public record OrCondition : Condition
{
    [JsonPropertyName("conditions")]
    public required List<Condition> Conditions { get; init; }
}

/// <summary>
/// Logical NOT condition.
/// </summary>
public record NotCondition : Condition
{
    [JsonPropertyName("condition")]
    public required Condition Cond { get; init; }
}

/// <summary>
/// Field existence condition.
/// </summary>
public record ExistsCondition : Condition
{
    [JsonPropertyName("field")]
    public required string Field { get; init; }
}

/// <summary>
/// Value in set condition.
/// </summary>
public record InCondition : Condition
{
    [JsonPropertyName("field")]
    public required string Field { get; init; }

    [JsonPropertyName("values")]
    public required List<object> Values { get; init; }
}

/// <summary>
/// Numeric range condition.
/// </summary>
public record RangeCondition : Condition
{
    [JsonPropertyName("field")]
    public required string Field { get; init; }

    [JsonPropertyName("min")]
    public double? Min { get; init; }

    [JsonPropertyName("max")]
    public double? Max { get; init; }

    [JsonPropertyName("min_inclusive")]
    public bool MinInclusive { get; init; } = true;

    [JsonPropertyName("max_inclusive")]
    public bool MaxInclusive { get; init; } = true;
}

/// <summary>
/// Custom expression condition.
/// </summary>
public record ExpressionCondition : Condition
{
    [JsonPropertyName("expression")]
    public required string Expression { get; init; }

    [JsonPropertyName("language")]
    public string Language { get; init; } = "cel";
}

File: arka-sdk-java/src/main/java/com/arka/sdk/types/Condition.java
package com.arka.sdk.types;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import java.util.List;

/**
 * Base class for rule conditions.
 */
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
@JsonSubTypes({
    @JsonSubTypes.Type(value = Condition.Compare.class, name = "compare"),
    @JsonSubTypes.Type(value = Condition.And.class, name = "and"),
    @JsonSubTypes.Type(value = Condition.Or.class, name = "or"),
    @JsonSubTypes.Type(value = Condition.Not.class, name = "not"),
    @JsonSubTypes.Type(value = Condition.Exists.class, name = "exists"),
    @JsonSubTypes.Type(value = Condition.In.class, name = "in"),
    @JsonSubTypes.Type(value = Condition.Range.class, name = "range"),
    @JsonSubTypes.Type(value = Condition.Expression.class, name = "expression")
})
public sealed interface Condition permits
    Condition.Compare, Condition.And, Condition.Or, Condition.Not,
    Condition.Exists, Condition.In, Condition.Range, Condition.Expression {

    record Compare(String field, String operator, Object value) implements Condition {}

    record And(List<Condition> conditions) implements Condition {}

    record Or(List<Condition> conditions) implements Condition {}

    record Not(Condition condition) implements Condition {}

    record Exists(String field) implements Condition {}

    record In(String field, List<Object> values) implements Condition {}

    record Range(
        String field,
        Double min,
        Double max,
        @JsonProperty("min_inclusive") boolean minInclusive,
        @JsonProperty("max_inclusive") boolean maxInclusive
    ) implements Condition {
        public Range {
            if (!minInclusive) minInclusive = true;
            if (!maxInclusive) maxInclusive = true;
        }
    }

    record Expression(String expression, String language) implements Condition {
        public Expression {
            if (language == null) language = "cel";
        }
    }
}
